nomad/stream/event_publisher.go

```diff
@@ -0,0 +1,222 @@
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	return e.eventBuf.Len()
+}
+
+// Publish events to all subscribers of the event Topic.
+func (e *EventBroker) Publish(events *structs.Events) {
+	if len(events.Events) > 0 {
+		e.publishCh <- events
+	}
+}
+
+// Subscribe returns a new Subscription for a given request. A Subscription
+// will receive an initial empty currentItem value which points to the first item
+// in the buffer. This allows the new subscription to call Next() without first checking
+// for the current Item.
+//
+// A Subscription will start at the requested index, or as close as possible to
+// the requested index if it is no longer in the buffer. If StartExactlyAtIndex is
+// set and the index is no longer in the buffer or not yet in the buffer an error
+// will be returned.
+//
+// When a caller is finished with the subscription it must call Subscription.Unsubscribe
+// to free ACL tracking resources. TODO(drew) ACL tracking
+func (e *EventBroker) Subscribe(req *SubscribeRequest) (*Subscription, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	var head *bufferItem
+	var offset int
+	if req.Index != 0 {
+		head, offset = e.eventBuf.StartAtClosest(req.Index)
+	} else {
+		head = e.eventBuf.Head()
+	}
+	if offset > 0 && req.StartExactlyAtIndex {
+		return nil, fmt.Errorf("requested index not in buffer")
+	} else if offset > 0 {
+		e.logger.Debug("requested index no longer in buffer", "requsted", int(req.Index), "closest", int(head.Events.Index))
+	}
+
+	// Empty head so that calling Next on sub
+	start := newBufferItem(&structs.Events{Index: req.Index})
+	start.link.next.Store(head)
+	close(start.link.ch)
+
+	sub := newSubscription(req, start, e.subscriptions.unsubscribeFn(req))
+
+	e.subscriptions.add(req, sub)
+	return sub, nil
+}
+
+// CloseAll closes all subscriptions
+func (e *EventBroker) CloseAll() {
+	e.subscriptions.closeAll()
+}
+
+func (e *EventBroker) handleUpdates(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			e.subscriptions.closeAll()
+			return
+		case update := <-e.publishCh:
+			e.sendEvents(update)
+		}
+	}
+}
+
+// sendEvents sends the given events to the publishers event buffer.
+func (e *EventBroker) sendEvents(update *structs.Events) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	e.eventBuf.Append(update)
+}
+
+type subscriptions struct {
+	// mu for byToken. If both subscription.mu and EventBroker.mu need
+	// to be held, EventBroker mutex MUST always be acquired first.
+	mu sync.RWMutex
+
+	// byToken is an mapping of active Subscriptions indexed by a token and
+	// a pointer to the request.
+	// When the token is modified all subscriptions under that token will be
+	// reloaded.
+	// A subscription may be unsubscribed by using the pointer to the request.
+	byToken map[string]map[*SubscribeRequest]*Subscription
+}
+
+func (s *subscriptions) add(req *SubscribeRequest, sub *Subscription) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	subsByToken, ok := s.byToken[req.Token]
```

Comment:
so the ACl token can still be blank and this works fine, since its a map[string]map[*SubscribeRequest]*Subscription, a new subscription is returned with an unsubscribe fn that has the pointer to the subscribe request so its fine for token to be an empty string.